### PR TITLE
Integration Tests

### PR DESCRIPTION
**Pull request by joshpalis**
- Integration Tests in progress but mostly complete, besides deserialization/ serialization.
- Updated Action Listener to include server timeout  parameter for testing considerations
- modified main script

### CODE REVIEW

**Comment by owaiskazi19** [comment]
Let's merge this after https://github.com/owaiskazi19/IndependentPlugin/pull/8.

File: src/test/java/ExtensionIT.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
+import org.junit.jupiter.api.BeforeEach;
```
Comment (owaiskazi19) [inline src/test/java/ExtensionIT.java]: Let's name this file as `TransportCommunicationIT.java`

File: src/test/java/ExtensionIT.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    // Create input/output stream to read/write to server
+                    BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                    PrintStream out = new PrintStream(socket.getOutputStream());
+
+                    // server logs MESSAGE RECIEVED: TESTTT
+                    // note : message validation is only done if message length >= 6 bytes
+                    // character is 1 byte
+                    out.print("TESTTT");
+
+                    // Exception will originate from org.opensearch.transport.TcpTransport
```
Comment (owaiskazi19) [inline src/test/java/ExtensionIT.java]: Change the single line comments to multi line comments if comments has multiple lines here and for the rest

File: src/test/java/ExtensionIT.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        // start transport service and attempt client connection
+        startTransportandClient(settings, client);
+
+        // expecting server response "Connection refused"
+        assertEquals("Connection refused", clientResult);
+    }
+
+    // test extension handshake recieved and acknowledged
+    @Test
+    public void testHandshakeRequestRecievedAndAcknowledged() {
```
Comment (owaiskazi19) [inline src/test/java/ExtensionIT.java]: Let's remove this method for now and add it once test is done

File: src/test/java/ExtensionIT.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                    Socket socket = new Socket(host, port);
+
+                    // Create input/output stream to read/write to server
+                    BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                    PrintStream out = new PrintStream(socket.getOutputStream());
+
+                    // server logs MESSAGE RECIEVED: TESTTT
+                    // note : message validation is only done if message length >= 6 bytes
+                    // character is 1 byte
+                    out.print("TESTTT");
```
Comment (owaiskazi19) [inline src/test/java/ExtensionIT.java]: Name it as `TEST`

Comment (joshpalis) [reply src/test/java/ExtensionIT.java]: TcpTransport executes message validation once the message length is 6 bytes. So I have to pass in at least 6 characters to trigger this behavior

Comment (owaiskazi19) [reply src/test/java/ExtensionIT.java]: Makes sense. Change it to `TRANSPORT_TEST` or something meaningful

File: src/test/java/ExtensionIT.java (hunk 2 above)
Comment (owaiskazi19) [reply src/test/java/ExtensionIT.java]: I see you have removed the comment. It's better to have comments, just change it to multi line comment

File: src/test/java/TransportCommunicationIT.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    // public void testHandshakeRequestRecievedAndAcknowledged() {
+
+    // send handshake request to SDK transport service
+    // - this is done when opensearch transport attempts to connect to extension node
+    // - SDK expected behavior is to recieve and validate message, and identify message as a handshake request
+    // - upon identification, SDK logs handshake request recieved
+    // - SDK then sends response to acknowledge handshake request
+    // assert that SDK logs : [internal:tcp/handshake] received request
+    // assert that SDK logs : [internal:tcp/handshake] sent response
+    // }
```
Comment (owaiskazi19) [inline src/test/java/TransportCommunicationIT.java]: Do we need this?

Comment (joshpalis) [reply src/test/java/TransportCommunicationIT.java]: I'll remove this for now

File: src/main/java/transportservice/ActionListener.java [hunk 6]
```diff
@@ -16,8 +16,12 @@ protected static InetSocketAddress getLocalEphemeral() throws UnknownHostExcepti
         return new InetSocketAddress(InetAddress.getLocalHost(), 0);
     }
 
-    public void runActionListener(boolean flag) {
+    public void runActionListener(boolean flag, int timeout) {
         try (ServerSocket socket = new ServerSocket()) {
+
+            // for testing considerations, otherwise zero which is interpreted as an infinite timeout
+            socket.setSoTimeout(timeout);
```
Comment (owaiskazi19) [inline src/main/java/transportservice/ActionListener.java:23]: Can we use the same timeout here?https://github.com/owaiskazi19/IndependentPlugin/blob/44bb779af05732c1326485267722e5f229286376/src/main/java/transportservice/ActionListener.java#L50

File: src/test/java/TransportCommunicationIT.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    private final int minPort = 49152;
+    private final int maxPort = 65535;
+    private final String host = "127.0.0.1";
+    private volatile String clientResult;
+
+    @BeforeEach
+    public void setUp() {
+
+        // randomize port selection when creating settings
+        port = getRandomPort();
```
Comment (owaiskazi19) [inline src/test/java/TransportCommunicationIT.java]: No need to create a randomPort here, which is actually within the range 49152-65535. Using an explicit port might not be free sometimes. Let's keep the test port as `0`.

**Comment by saratvemulapalli** [comment]
Looks like gradle check is failing.
@owaiskazi19 how much of a work is it to get gradle check running?

```
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not find org.opensearch:opensearch:2.0.0-SNAPSHOT.
     Searched in the following locations:
       - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
       - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
       - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
     Required by:
         project :
```

File: src/test/java/TransportCommunicationIT.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            for (TransportAddress transportAddress : transport.boundAddress().boundAddresses()) {
+                assert (transportAddress instanceof TransportAddress);
+                assertEquals(host, transportAddress.getAddress());
+                assertEquals(port, transportAddress.getPort());
+            }
+
+            // check publish addresses
+            assert (transport.boundAddress().publishAddress() instanceof TransportAddress);
+            TransportAddress publishAddress = transport.boundAddress().publishAddress();
+            assertEquals(host, NetworkAddress.format(publishAddress.address().getAddress()));
```
Comment (owaiskazi19) [inline src/test/java/TransportCommunicationIT.java:61]: assertEquals syntax is assertEquals(expected, actual). Need to swap the arguments here and for the rest

**Review by owaiskazi19 — CHANGES_REQUESTED** [review]
assertEquals syntax is assertEquals(expected, actual). Need to swap the arguments here and for the rest

File: src/main/java/transportservice/RunPlugin.java [hunk 9]
```diff
@@ -41,22 +41,31 @@
...
+
+        NetworkService networkService = new NetworkService(Collections.emptyList());
+        PageCacheRecycler pageCacheRecycler = new PageCacheRecycler(settings);
+        IndicesModule indicesModule = new IndicesModule(Collections.emptyList());
+        SearchModule searchModule = new SearchModule(settings, Collections.emptyList());
+
+        List<NamedWriteableRegistry.Entry> namedWriteables = Stream.of(
+            NetworkModule.getNamedWriteables().stream(),
+            indicesModule.getNamedWriteables().stream(),
+            searchModule.getNamedWriteables().stream(),
```
Comment (saratvemulapalli) [inline src/main/java/transportservice/RunPlugin.java:61]: Is this code from the merge?

These modules wouldn't exist on the extension. 
Can we have a non null Empty NamedWritables instead?

**Comment by owaiskazi19** [comment]
> Looks like gradle check is failing. @owaiskazi19 how much of a work is it to get gradle check running?
> 
> ```
> > Could not resolve all files for configuration ':compileClasspath'.
>    > Could not find org.opensearch:opensearch:2.0.0-SNAPSHOT.
>      Searched in the following locations:
>        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
>        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
>        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
>        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
>      Required by:
>          project :
> ```

Hey @saratvemulapalli! Gradle check will fail because maven dependencies are getting fetched from local and not from maven central. We need to publish the code from the feature branch to maven central and it will resolve the gradle check failure.

**Comment by saratvemulapalli** [comment]
> > Looks like gradle check is failing. @owaiskazi19 how much of a work is it to get gradle check running?
> > ```
> > > Could not resolve all files for configuration ':compileClasspath'.
> >    > Could not find org.opensearch:opensearch:2.0.0-SNAPSHOT.
> >      Searched in the following locations:
> >        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
> >        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
> >        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
> >        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
> >      Required by:
> >          project :
> > ```
> 
> Hey @saratvemulapalli! Gradle check will fail because maven dependencies are getting fetched from local and not from maven central. We need to publish the code from the feature branch to maven central and it will resolve the gradle check failure.

Could you open an issue? one of us can pick it up. 
We should be able to publish OpenSearch to maven local in the github CI.

**Comment by owaiskazi19** [comment]
> > > Looks like gradle check is failing. @owaiskazi19 how much of a work is it to get gradle check running?
> > > ```
> > > > Could not resolve all files for configuration ':compileClasspath'.
> > >    > Could not find org.opensearch:opensearch:2.0.0-SNAPSHOT.
> > >      Searched in the following locations:
> > >        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
> > >        - file:/home/runner/.m2/repository/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
> > >        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/maven-metadata.xml
> > >        - https://repo.maven.apache.org/maven2/org/opensearch/opensearch/2.0.0-SNAPSHOT/opensearch-2.0.0-SNAPSHOT.pom
> > >      Required by:
> > >          project :
> > > ```
> > 
> > 
> > Hey @saratvemulapalli! Gradle check will fail because maven dependencies are getting fetched from local and not from maven central. We need to publish the code from the feature branch to maven central and it will resolve the gradle check failure.
> 
> Could you open an issue? one of us can pick it up. We should be able to publish OpenSearch to maven local in the github CI.

https://github.com/opensearch-project/OpenSearch/issues/2758. This is a great idea to publish the dependencies to maven local in GHA.

File: src/test/java/TransportCommunicationIT.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+        settings = Settings.builder()
+            .put("node.name", "node_extension_test")
+            .put(TransportSettings.BIND_HOST.getKey(), host)
+            .put(TransportSettings.PORT.getKey(), port)
+            .build();
+    }
+
+    @Test
+    public void testThatInfosAreExposed() {
```
Comment (saratvemulapalli) [inline src/test/java/TransportCommunicationIT.java]: super nit: This sounds funny, may be `testSocketSetup` ?

File: src/test/java/TransportCommunicationIT.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                assertEquals(port, transportAddress.getPort());
+            }
+
+            // check publish addresses
+            assert (transport.boundAddress().publishAddress() instanceof TransportAddress);
+            TransportAddress publishAddress = transport.boundAddress().publishAddress();
+            assertEquals(host, NetworkAddress.format(publishAddress.address().getAddress()));
+            assertEquals(port, publishAddress.address().getPort());
+
+        } finally {
```
Comment (saratvemulapalli) [inline src/test/java/TransportCommunicationIT.java:60]: Shouldn't we terminate the socket? 
Wouldn't this lead to lingering socket pairs?

**Review by saratvemulapalli — APPROVED** [review]
Thanks @joshpalis for working on this. 
This is really nice!

Couple of questions:
1. Are these tests run as part of gradle check?
2. Could we write how to invoke these tests in our [documentation](https://github.com/owaiskazi19/IndependentPlugin/blob/main/DEVELOPER_GUIDE.md#run-tests)?
Dropped couple of other minor comments:

File: src/test/java/TransportCommunicationIT.java (hunk 10 above)
Comment (joshpalis) [reply src/test/java/TransportCommunicationIT.java]: sure ill change the name

File: src/main/java/transportservice/RunPlugin.java (hunk 9 above)
Comment (joshpalis) [reply src/main/java/transportservice/RunPlugin.java:61]: Yes this code is from merging the current main branch

File: src/test/java/TransportCommunicationIT.java (hunk 11 above)
Comment (joshpalis) [reply src/test/java/TransportCommunicationIT.java:60]: Thanks for the suggestion, I'll add this change in my next commit

**Review by owaiskazi19 — APPROVED** [review]
Thanks for the tests. Can we add a CI job for tests?

**Comment by owaiskazi19** [comment]
Rebase the branch with `main` to run the latest CI for gradle check

**Comment by owaiskazi19** [comment]
@joshpalis gradle check failure

```
TransportCommunicationIT > testInvalidMessageFormat() STANDARD_ERROR
    Exception in thread "Thread-5" 
    java.io.UncheckedIOException: java.net.SocketTimeoutException: Accept timed out

    	at transportservice.ActionListener$1.run(ActionListener.java:36)

    Caused by: java.net.SocketTimeoutException: Accept timed out

    	at java.base/java.net.PlainSocketImpl.socketAccept(Native Method)

    	at java.base/java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:458)

    	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:565)

    	at java.base/java.net.ServerSocket.accept(ServerSocket.java:[53](https://github.com/owaiskazi19/IndependentPlugin/runs/5826049630?check_suite_focus=true#step:6:53)3)

    	at transportservice.ActionListener$1.run(ActionListener.java:31)


TransportCommunicationIT > testInvalidMessageFormat() PASSED

TransportCommunicationIT > testMismatchingPort() STANDARD_OUT
    [2022-04-05T00:42:37,391][INFO ][o.o.t.TransportService   ] [[Test worker]] Registered handler with actioninternal:transport/handshake
    [2022-04-05T00:42:37,397][INFO ][o.o.t.TransportService   ] [[Test worker]] publish_address {127.0.0.1:9301}, bound_addresses {127.0.0.1:9301}
    [2022-04-05T00:42:37,398][INFO ][o.o.t.TransportService   ] [[Test worker]] Registered handler with actioninternal:discovery/extensions

TransportCommunicationIT > testMismatchingPort() STANDARD_ERROR
    Exception in thread "Thread-7" 
    java.io.UncheckedIOException: java.net.SocketTimeoutException: Accept timed out

    	at transportservice.ActionListener$1.run(ActionListener.java:36)

    Caused by: java.net.SocketTimeoutException: Accept timed out

    	at java.base/java.net.PlainSocketImpl.socketAccept(Native Method)

    	at java.base/java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:458)

    	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:[56](https://github.com/owaiskazi19/IndependentPlugin/runs/5826049630?check_suite_focus=true#step:6:56)5)

    	at java.base/java.net.ServerSocket.accept(ServerSocket.java:533)

    	at transportservice.ActionListener$1.run(ActionListener.java:31)


TransportCommunicationIT > testMismatchingPort() FAILED
    org.junit.ComparisonFailure at TransportCommunicationIT.java:130

**Pull request merged**